README.md
WellRESTed
==========

WellRESTed provides classes to help you create RESTful APIs and work with HTTP requests and responses.


Requirements
------------

- PHP 5.3
- [Composer](http://getcomposer.org/) for autoloading
- [PHP cURL](http://php.net/manual/en/book.curl.php) for making requests


Install
-------

Add an entry for "pjdietz/wellrested" to your composer.json file's **require** property. If you are not already using Composer, create a file in your project called **composer.json** with the following content:

```json
{
    "require": {
        "pjdietz/wellrested": "1.*"
    }
}
```

Use Composer to download and install WellRESTed. Run these commands from the directory containing the **composer.json** file.

```bash
$ curl -s https://getcomposer.org/installer | php
$ php composer.phar install
```

You can now use WellRESTed by including the **autoload.php** file generated by Composer. `vendor/autoload.php`


Examples
--------

### Routing

WellRESTed's primary goal is to facilitate mapping of URIs to classes that will provide or accept representations. To do this, create a Router instance and load it up with some Routes. Each Route is simply a mapping of a URI pattern to a class name. The class name represents the Handler class which the router will dispatch at the time it receives a request for the given URI. The handlers are never loaded unless they are needed.

Here's an example of a Router that will handle two URIs:

```php
// Build the router.
$myRouter = new Router();
$myRouter->addRoute(Route::newFromUriTemplate('/things/', 'ThingCollectionHandler'));
$myRouter->addRoute(Route::newFromUriTemplate('/things/{id}', 'ThingItemHandler'));

// Determine and output the response.
$response = $myRouter->getResponse();
$response->respond();
```

When you create your Handler subclass, you will provide a method for each HTTP verb you would like the endpoint to support. For example, if `/things/` should support `GET`, you would override the `get()` method. For `POST`, `post()`, etc.

If your endpoint should reject particular verbs, no worries. The Handler base class defines the default verb-handling methods to respond with a **405 Method Not Allowed** status.

Here's a simple Handler that matches the first endpoint, `/things/`.

```php
class ThingCollectionHandler extends \pjdietz\WellRESTed\Handler
{
    protected function get()
    {
        // Read some things from the database, cache, whatever.
        // ...read this into the variable $data

        // Set the values for the instance's response member. This is what the
        // Router will eventually use to output a response to the client.
        $this->response->setStatusCode(200);
        $this->response->setHeader('Content-Type', 'application/json');
        $this->response->setBody(json_encode($data));
    }

    protected function post()
    {
        // Read from the instance's request member and store a new Thing.
        // ...

        // Build a response to send to the client.
        $this->response->setStatusCode(201);
        $this->response->setBody('You added a thing!');
    }
}
```

This Handler works with the second endpoint, `/things/{id}`. The pattern for this endpoint has the variable `{id}` in it. The Handler can access path variables through its `args` member, which is an associative array of variables from the URI.

```php
class ThingItemHandler extends \pjdietz\WellRESTed\Handler
{
    protected function get()
    {
        // Lookup a Thing ($thing) based on $this->args['id']
        // ...

        if ($thing) {
            // The Thing exists! Let's output a representation.
            $this->response->setStatusCode(200);
            $this->response->setHeader('Content-Type', 'application/json');
            $this->response->setBody(json_encode($thing));
        } else {
            // The ID did not match anything.
            $this->response->setStatusCode(404);
            $this->response->setHeader('Content-Type', 'text/plain');
            $this->response->setBody('No thing with id ' . $this->args['id']);
        }
    }
}
```

As of version 1.3.0, you can have a Router dispatch other Routers. This is useful for breaking your endpoints into smaller route tables. Here's an example:

```php
/**
 * Top-level router that dispatches all traffic to /cat/... to CatRouter and all
 * traffic to /dog/... to DogRouter.
 */
class TopRouter extends \pjdietz\WellRESTed\Router
{
    public function __construct()
    {
        $this->addRoute(new Route('/^\/cat\//', 'CatRouter'));
        $this->addRoute(new Route('/^\/dog\//', 'DogRouter'));
    }
}
```


### Requests and Responses

You've already seen a Response in use in the examples above. You can also use Responses outside of Handlers. Let's take a look at creating a new Response, setting a header, supplying the body, and outputting.

```php
$resp = new \pjdietz\WellRESTed\Response();
$resp->setStatusCode(200);
$resp->setHeader('Content-type', 'text/plain');
$resp->setBody('Hello world!');
$resp->respond();
exit;
```

The Request class goes hand-in-hand with the Response class. Again, this is used in the Handler class to read the information from the request being handled. From outside the context of a Handler, you can also use the Request class to read info for the request sent to the server.

```php
// Call the static method getRequest() to get a reference to the Request
// singleton that represents the request made to the server.
$rqst = \pjdietz\WellRESTed\Request::getRequest();

if ($rqst->getMethod() === 'PUT') {
    $obj = json_decode($rqst->getBody());
    // Do something with the JSON sent as the message body.
    // ...
}
```

The Request class can also make a request to another server and provide the response as a Response object. (This feature requires [PHP cURL](http://php.net/manual/en/book.curl.php).)

```php
// Prepare a request.
$rqst = new \pjdietz\WellRESTed\Request();
$rqst->setUri('http://my.api.local/resources/');
$rqst->setMethod('POST');
$rqst->setBody(json_encode($newResource));

// Make the request.
$resp = $rqst->request();

// Read the response.
if ($resp->getStatusCode() === 201) {
    // The new resource was created.
    $createdResource = json_decode($resp->getBody());
}
```


### Routing from a Handler

One more fun thing you can do with WellRESTed is use your API from inside your API. WellRESTed makes this easy because each dispatched Handler or Router keeps a reference to the top-level Router.

Suppose you have an endpoint that needs to look up information using another endpoint. Now that you've seen that you can create and work with your own requests and responses, we'll look at how to use them in the context of a handler.

```php
class NestedHandler extends \pjdietz\WellRESTed\Handler
{
    protected function get()
    {
        // To build this response, we need some of the info from the /things/ representation.
        // Create a Request instance to query this API.
        $rqst = new Request();
        $rqst->setPath('/things/');

        // Use this handler's reference to the router that dispatched it.
        $router = $this->getRouter();

        // Pass the request for /things/ to the router to get a response.
        $resp = $router->getResponse($rqst);

        // Read the status code, headers, body, etc. on the response.
        // ...
    }
}
```


More Examples
---------------

For more examples, see the project [pjdietz/wellrested-samples](https://github.com/pjdietz/wellrested-samples).


Copyright and License
---------------------
Copyright © 2013 by PJ Dietz
Licensed under the [MIT license](http://opensource.org/licenses/MIT)


CHANGES.md
### v1.3.0
- Allow Routers to dispatch additional Routers.

### v1.2.2
- Fix issue with 100 Continue status codes, etc. where r\n\r\n appears within the list of headers

### v1.2.1
- Allow Request to read request headers without apache_request_headers().

### v1.2.0
2013-05-27
**Warning** Some of the changes in this update will break backward compatibility.
- Remove constructor from Handler.
- Add Interfaces: RouterInterface, RouteInterface, HandlerInterface, RequestInterface, and ResponseInterface
- Classes are updated to implement interfaces and type hint interfaces in method calls. For example, Router now implements RouteInterface and its method addRoute() now type hints the parameter as a RouteInterface instead of a Route.
- Route fields pattern and handler are now private and replaced accessors to conform to RouteInterface
- Add support for port in Request. You can now use Request::setPort() and Request::getPort(), and the port is included and extract from the URI.

### v1.1.2
2013-05-19
- Change Handler to an abstract class
- Add Handler::respondWithMethodNotAllowed() for flexibility
- Convert a number of methods from protected to private
- Bug fix: return null or false from magic __get() and __isset() methods

### v1.1.1
2013-03-29
- Bug fix: Instantiate Message->headers and Message->headerLines in constructor

### v1.1.0
2013-03-26
- Add MIT License to composer.json
- Add constructor to Request to allow setting URI and method on instantiation
- Add Response::getSuccess()


Response.inc.php
<?php

namespace pjdietz\WellRESTed;

require_once(dirname(__FILE__) . '/Message.inc.php');

/*******************************************************************************
 * Response
 *
 * A Response instance allows you to build an HTTP response and send it when
 * finished.
 *
 * @package WellRESTed
 *
 ******************************************************************************/

/**
 * @property string reasonPhrase  Text explanation of status code.
 * @property int statusCode  HTTP status code
 * @property string statusLine  HTTP status line, e.g. "HTTP/1.1 200 OK"
 */
class Response extends Message {

    /**
     * Text explanation of the HTTP Status Code. You only need to set this if
     * you are using nonstandard status codes. Otherwise, the instance will
     * set the when you update the status code.
     *
     * @var string
     */
    protected $reasonPhrase;

    /**
     * HTTP status code
     * @var int
     */
    protected $statusCode;

    // -------------------------------------------------------------------------

    /**
     * Create a new Response instance, optionally passing a status code, body,
     * and headers.
     *
     * @param int $statusCode
     * @param string $body
     * @param array $headers
     */
    public function __construct($statusCode=500, $body=null, $headers=null) {

        $this->statusCode = $statusCode;

        if (is_array($headers)) {
            $this->headers = $headers;
        } else {
             $this->headers = array();
        }

        if (!is_null($body)) {
            $this->body = $body;
        }

        if (isset($_SERVER['SERVER_PROTOCOL'])) {
            $this->protocol = $_SERVER['SERVER_PROTOCOL'];
        } else {
            $this->protocol = 'HTTP/1.1';
        }

    }

    // -------------------------------------------------------------------------
    // !Accessors

    /**
     * @param string $name
     * @return array|string
     * @throws \Exception
     */
    public function __get($name) {

        switch ($name) {
            case 'reasonPhrase':
                return $this->getReasonPhrase();
            case 'statusCode':
                return $this->getStatusCode();
            case 'statusLine':
                return $this->getStatusLine();
            default:
                return parent::__get($name);
        }

    }

    /**
     * @param string $name
     * @param mixed $value
     * @throws \Exception
     */
    public function __set($name, $value) {

        switch ($name) {
            case 'reasonPhrase':
                $this->setReasonPhrase($value);
                return;
            case 'statusCode':
                $this->setStatusCode($value);
                return;
            default:
                parent::__set($name, $value);
        }

    }

    /**
     * Provide a new entity body for the respone.
     * This method also updates the content-length header based on the length
     * of the new body string.
     *
     * @param string $value
     * @param bool $setContentLength  Automatically add a Content-length header
     */
    public function setBody($value, $setContentLength=true) {

        $this->body = $value;

        if ($setContentLength === true) {
            $this->setHeader('Content-Length', strlen($value));
        }

    }

    /**
     * @return string
     */
    public function getReasonPhrase() {
        return $this->reasonPhrase;
    }

    /**
     * @param string $statusCodeMessage
     */
    public function setReasonPhrase($statusCodeMessage) {
        $this->reasonPhrase = $statusCodeMessage;
    }

    /**
     * @return int
     */
    public function getStatusCode() {
        return $this->statusCode;
    }

    /**
     * @param int $statusCode
     * @param string $reasonPhrase
     * @throws \InvalidArgumentException
     * @return void
     */
    public function setStatusCode($statusCode, $reasonPhrase=null) {

        $this->statusCode = (int) $statusCode;

        if (is_null($reasonPhrase)) {

            switch ($this->statusCode) {
                case 100: $text = 'Continue'; break;
                case 101: $text = 'Switching Protocols'; break;
                case 200: $text = 'OK'; break;
                case 201: $text = 'Created'; break;
                case 202: $text = 'Accepted'; break;
                case 203: $text = 'Non-Authoritative Information'; break;
                case 204: $text = 'No Content'; break;
                case 205: $text = 'Reset Content'; break;
                case 206: $text = 'Partial Content'; break;
                case 300: $text = 'Multiple Choices'; break;
                case 301: $text = 'Moved Permanently'; break;
                case 302: $text = 'Moved Temporarily'; break;
                case 303: $text = 'See Other'; break;
                case 304: $text = 'Not Modified'; break;
                case 305: $text = 'Use Proxy'; break;
                case 400: $text = 'Bad Request'; break;
                case 401: $text = 'Unauthorized'; break;
                case 402: $text = 'Payment Required'; break;
                case 403: $text = 'Forbidden'; break;
                case 404: $text = 'Not Found'; break;
                case 405: $text = 'Method Not Allowed'; break;
                case 406: $text = 'Not Acceptable'; break;
                case 407: $text = 'Proxy Authentication Required'; break;
                case 408: $text = 'Request Time-out'; break;
                case 409: $text = 'Conflict'; break;
                case 410: $text = 'Gone'; break;
                case 411: $text = 'Length Required'; break;
                case 412: $text = 'Precondition Failed'; break;
                case 413: $text = 'Request Entity Too Large'; break;
                case 414: $text = 'Request-URI Too Large'; break;
                case 415: $text = 'Unsupported Media Type'; break;
                case 500: $text = 'Internal Server Error'; break;
                case 501: $text = 'Not Implemented'; break;
                case 502: $text = 'Bad Gateway'; break;
                case 503: $text = 'Service Unavailable'; break;
                case 504: $text = 'Gateway Time-out'; break;
                case 505: $text = 'HTTP Version not supported'; break;
                default:  $text = 'Nonstandard'; break;
            }

            $this->reasonPhrase = $text;

        } else {

            if (is_string($reasonPhrase)) {
                $this->reasonPhrase = $reasonPhrase;
            } else {
                throw new \InvalidArgumentException('$reasonPhrase must be a string (or null to use standard HTTP Reason-Phrase');
            }

        }

    }

    /**
     * Return HTTP status line, e.g. HTTP/1.1 200 OK.
     *
     * @return string
     */
    protected function getStatusLine() {
        return sprintf('%s/%s %s %s',
                    strtoupper($this->protocol),
                    $this->protocolVersion,
                    $this->statusCode,
                    $this->reasonPhrase);
    }

    // -------------------------------------------------------------------------

    /**
     * Output the response to the client.
     *
     * @param bool $headersOnly  Do not include the body, only the headers.
     */
    public function respond($headersOnly=false) {

        // Output the HTTP status code.
        header($this->statusLine);

        // Output each header.
        foreach ($this->headers as $header => $value) {
            header($header . ': ' . $value);
        }

        // Output the entity body.
        if (!$headersOnly && isset($this->body)) {
            print $this->body;
        }

    }

}

?>
